### ISSUE TITLE
Muting PulseEffects Source does nothing.

### ISSUE THREAD

**Issue by kevincox** [wwmm/easyeffects#956]
## Steps to Reproduce

1. Open pavucontrol.
2. Mute PulseEffects Source.
3. Speak into your microphone.
4. Notice that sound is still passed through PulseEffecs Source.

## Example Use Case

One notable use case that is broken here is that GNOME 3's "Microphone mute/unmute" option mutes the default input device. If you are using PulseEffects this does nothing which is inconvenient and surprising.

It seems reasonable that a lot of tools would mute the default input device, so it would be better if it did something useful.

## Workaround

Muting the input device that PulseEffects is recording from works as expected. (ex: `pactl set-source-mute alsa_input.usb-0000-My_Mic.mono-fallback toggle`)

**Comment by wwmm** [comment]
This is one of those situations where I am not sure if the real fix should be done on our side. In PipeWire it is totally fine to have virtual input and output devices without volume and mute controls. The problem is that the current audio managers like Pavucontrol and Gnome settings are still using Pulseaudio's API where this kind of thing does not exist. Every Pulseaudio's virtual device has volume and volume control so the managers always expose these widgets to the user. This approach was not a problem in Pulseaudio but it is in PipeWire.

The reason why I am disabling these controls is that this was source of headaches in Pulseaudio where I could not disable them. From time to time a user opened an issue here reporting that there was no sound. After lots of time wasted in investigation we eventually realized that something in the user system decided it was a good idea to mute our devices or setting their volumes to zero without telling the user this was being done. This has also happened to me in rare occasions. Sometimes when I opened the game Rise of the Tomb raider our devices suddenly had their volumes set to zero. I could never figure out if the bug was in Pulseaudio or in the game. So now I can disable these controls I do it.

I realize that it is weird to the user having controls that do not work. But so it is having third party tools or bugs in the audio server muting devices that are vital to us without letting the user aware of that. In my opinion the real fix for this situation is moving the audio managers to the PipeWire API. This way they will have a way to identify if it makes sense or not to show mute and volume controls depending on the type of device they are handling.

>It seems reasonable that a lot of tools would mute the default input device

In general I do not recommend that our virtual devices are set as default. If I had a way to disable that I would have done too a long time ago. In order to apply effects for microphone we have to know which mic the user is using. The default operation is asking the server default input device name when we initialize. For reasons that are not clear both in Pulseaudio and in PipeWire there is a risk we get our own device name instead of the user mic if the user has set our device as default. Even if we did not load it yet. So I do not recommend users do that.

**Comment by kevincox** [comment]
> In PipeWire it is totally fine to have virtual input and output devices without volume and mute controls. 

Interesting, I was not aware of that.

> The reason why I am disabling these controls is...

On the other hand this is a useful feature. It means that I can mute PulseEffects independently of my microphone. (Although maybe this is a niche use case).

But more interestingly is there a way in PipeWire to find the source and mute it? How *should* gnome implement this feature such that it "just works" in the common case as it does without PulseEffects. (Applications will record from the default input unless configured otherwise and GNOME will mute the default input). Maybe if there was some way to "delegate" muting to the source, but I suspect that will trigger a lot of bugs too.

> In general I do not recommend that our virtual devices are set as default.

Then how should I use PulseEffects for applications that don't support overriding the default audio device other than manually adjusting it every time? For example even on the web in Firefox some sites allow me to choose, some use the default and remember changes, some reset to the default every time and some sites don't let you choose anything but the default. I have a fairly loud keyboard and fairly low requirements on audio quality so *for me* using the noise canceled device is a reasonable default.

>  For reasons that are not clear both in Pulseaudio and in PipeWire there is a risk we get our own device name instead of the user mic if the user has set our device as default.

Yes, I have seen this with PulseEffects 5.0.2 and earlier. However I haven't seen it yet in 5.0.3. However this seems more like a bug than a compelling argument that I shouldn't change the default. (although I admit that other than asking the user there isn't a clear answer what device to use in this case, although if the device could be resolved before adding the PulseEffects Source than I think rebinding to that device if you are moved to your own source should solve it) Furthermore I would still rather fix which device PulseEffects is using once a day than adjusting it for every program that I open throughout the day.

**Comment by wwmm** [comment]
>How should gnome implement this feature such that it "just works" in the common case as it does without PulseEffects

In order to disable mute and volume controls I create the virtual device with the flag `monitor.channel-volumes = "false"`. It is really easy to check that. Not only with PipeWire's API but from the command line too. Run `pw-cli list-objects Node` to find the nodes and after identifying the desired device `id number` run `pw-cli info id_number`. If `monitor.channel-volumes` was set it will be in the output.

>Then how should I use PulseEffects for applications that don't support overriding the default audio device other than manually adjusting it every time?

I am not aware of another way that does not involve explaining to these applications developers that they are doing something wrong. Unfortunately.

>However this seems more like a bug than a compelling argument that I shouldn't change the default

It is a bug that is out of my reach. I can not fix whatever application is messing with other peoples volume and mute settings.

>although if the device could be resolved before adding the PulseEffects Source

That is not the problem. I ask the server about the default input device before trying to load ours. It seems to me that sometimes the server sends a cached value under some circumstances.

>I think rebinding to that device if you are moved to your own source should solve it

It is not that easy. If we receive our own device as default we already ignore it. But in the end we will have no input device to use. There is no device "to rebind to" because the server did not send the real input device used by the user. What we would have to do in this situation is iterating over all the inputs and choosing one hoping that it is the one the user is actually using.

>I would still rather fix which device PulseEffects is using once a day than adjusting it for every program that I open throughout the day.

I do not see why this would be necessary. If you want PulseEffects to apply effects to all these programs enable `Process All Inputs` or `Process All Outputs` in our settings menu. PulseEffects will move the applications to our virtual source automatically. That is what I do. I have never had the need to set our devices as default or manually moving programs to them.

**Comment by wwmm** [comment]
>I do not see why this would be necessary

The case where the application do not allow its stream to be moved is the only one where setting our devices as default is the only solution. For the other cases it should be enough to enable effects for all inputs or outputs.

**Comment by kevincox** [comment]
> In order to disable mute and volume controls

I mean how could someone provide an easy mute button that disables the audio input that I most likely want disabled. "Mute the default input device" seems like the obvious answer but you are suggesting that you don't want to let PulseEffects to be muted which is a problem when it is the default input device..

> I am not aware of another way that does not involve explaining to these applications developers that they are doing something wrong. Unfortunately.

I don't share your view that they are doing something wrong. They are using the default input device. That is what I expect them to do! In my mind they are doing the right thing. Why would I expect them to know to use a device that I haven't marked as the default by default?

> I can not fix whatever application is messing with other peoples volume and mute settings.

I was talking about getting switched to recording from your own input. This only involves PipeWire and PulseEffects IIUC.

> It seems to me that sometimes the server sends a cached value under some circumstances.

Ah I see. The server is sending you a device that doesn't (yet) exist? Have you reported this upstream? Or does it send you no device at all?

> PulseEffects will move the applications to our virtual devices automatically.

> The case where the application do not allow its stream to be moved is the only one where setting our devices as default is the only solution

I'm guessing that Firefox and Chromium fall into this category. Unfortunately that is the vast majority of my microphone use is via a web browser so for me setting the default is quite essential.

----

My understanding so far. We basically have 3 "issues" which compound to create this problem. Solving any one should resolve the whole issue.

- PulseEffects was designed to grab streams itself, not relying on the "Default" concept in PulseAudio/PipeWire.
- This interacts poorly with some applications (Notably Firefox and Chromium) which don't allow their streams to be moved.
- This necessitates setting PulseEffects as the default input device which breaks GNOME's "Mute Microphone" feature because PulseEffects Source can't be muted.

**Comment by wwmm** [comment]
>I don't share your view that they are doing something wrong. They are using the default input device. That is what I expect them to do! In my mind they are doing the right thing. Why would I expect them to know to use a device that I haven't marked as the default by default?

I get the feeling we are not understanding each other. What I meant is that they should not be forcing their audio stream to use only the default device. In other words these developers are setting a flag that does not allow their audio stream to be moved to another device. There is obviously no issue in choosing the default device initially. The problem is not allowing the initial selection to be changed. This is something I talked about to PipeWire's developer a few months ago. And he explained that the only valid reason for use such a flag has nothing to do with these applications use case. Some people really misunderstood the true purpose of the `Don't Move` Pulseaudio's flag...

>The server is sending you a device that doesn't (yet) exist? Have you reported this upstream? Or does it send you no device at all?

I do not remember if I reported or not. This can happen both in Pulseaudio and in PipeWire.

>I was talking about getting switched to recording from your own input. This only involves PipeWire and PulseEffects IIUC.

I do not understand what you mean by that.

>PulseEffects was designed to grab streams itself, not relying on the "Default" concept in PulseAudio/PipeWire.

Not entirely correct. We listen to default input and output devices changes and we use these devices if the user is not selecting a custom device. We also ignore when the server sends our own devices in the callback. The only problem is when the server does that when we are initializing and asking for the default values for the first time. That is when things may break because the variables that should have the default devices won't point to a valid device.

>This interacts poorly with some applications (Notably Firefox and Chromium) which don't allow their streams to be moved.

I do not remember having problems with Chromium. I use it regularly and so far the streams are moved. Firefox unfortunately does that. At least Firefox it allows the input device to be manually selected in its own interface. There are other applications that do not allow even this.

>This necessitates setting PulseEffects as the default input device which breaks GNOME's "Mute Microphone" feature because PulseEffects Source can't be muted.

That is right. It is not nice but if the real mic is muted the whole processing chain will be muted. And as a result our virtual source too.

**Comment by kevincox** [comment]
> What I meant is that they should not be forcing their audio stream to use only the default device.

I agree with that.

> I do not understand what you mean by that.

I'm talking about PulseEffects sometimes getting its own stream when you are recording.

> I do not remember having problems with Chromium. 

I was pretty sure I tested it there. But it isn't my primary browser so maybe I remember wrong or it has changed.

> At least Firefox it allows the input device to be manually selected in its own interface.

I don't see this. I get it the first time I approve the Mic on a site but once that is gone I don't seem to have the option to ever change it. I filed https://bugzilla.mozilla.org/show_bug.cgi?id=1708737 to see if they want to change their mind about disallowing moves.

**Comment by wwmm** [comment]
>I'm talking about PulseEffects sometimes getting its own stream when you are recording.

I do not remember that particular bug. But it is the kind of thing that won't happen anymore once the work in our gtk4 branch is finished. With PipeWire filters replacing GStreamer we won't have a stream that could be moved anymore.

>I was pretty sure I tested it there. But it isn't my primary browser so maybe I remember wrong or it has changed.

Firefox is also my primary browser. But the educational institution where I work is using Microsoft Teams for our online classes and that does not work in Firefox. And as the Teams app for Linux is not doing screen sharing on Wayland I am forced to give classes in Chromium... No problems moving its streams.

> I don't see this. I get it the first time I approve the Mic on a site but once that is gone I don't seem to have the option to ever change it. 

Maybe it is a one time selection. I do not remember well now.

>I filed https://bugzilla.mozilla.org/show_bug.cgi?id=1708737 to see if they want to change their mind about disallowing moves.

It does not look like they care about this problem. Look at a much older issue that someone else has opened about this https://bugzilla.mozilla.org/show_bug.cgi?id=1645684. PulseEffects is not the only one having problems with this decision. I even commented there but so far no response from Firefox developers...

I will enable these controls again in the gtk4 branch. But I bet that the "I have no sound" issues caused by muted devices will keep coming =/

**Comment by kevincox** [comment]
> No problems moving [chromium's] streams.

You are right, I just confirmed this.

> Take a look at an older issue that someone else has opened about this https://bugzilla.mozilla.org/show_bug.cgi?id=1645684

Ah, I couldn't find that one. I'll close mine.

> I will enable these controls again in the gtk4 branch.

Thank you! Hopefully is isn't too bad :( It seems like everything just quite doesn't fit together here at the moment unfortunately.

**Comment by wwmm** [comment]
I think that PipeWire is disabling these controls by default now. Even after removing from the gtk4 branch the flag I was setting  these controls still do nothing. So it may be worth to make a question in Pipewire's page at Gitlab.

**Comment by kevincox** [comment]
What exactly is the question? If mute should be supported by default? I'm not sure I have enough context to ask as I have never looked at the pipewire APIs.

**Comment by wwmm** [comment]
It is not necessary anymore. After looking another issue there I found out that now I have to explicitly enable these controls for them to work.

**Comment by wwmm** [comment]
This should be fixed in EasyEffects 6.0.0

**Comment by sim590** [comment]
I'm on 7.1.8 and this is still not working. Muting `EasyEffects Source` in pulsemixer for instance:
![image](https://github.com/user-attachments/assets/eb24daff-12b9-41ab-8c1a-7a11543f061d)
still lets the sound be passed to application using that source. For instance, in this capture, Chromium Input is Discord and the sound is indeed going through. Following @wwmm's last message, I was under the impression that this was fixed in 6.0.0. What am I missing?

**Comment by wwmm** [comment]
> I'm on 7.1.8 and this is still not working. Muting `Easy Effects Source` in pulsemixer for instance:
> ![image](https://github.com/user-attachments/assets/eb24daff-12b9-41ab-8c1a-7a11543f061d)
> still lets the sound be passed to application using that source. For instance, in this capture, `Chromium input` is Discord and the sound is indeed going through. Following @wwmm's last message, I was under the impression that this was fixed in 6.0.0. What am I missing? 

Nowadays we just follow pipewire's default behavior. So mute and volume controls of virtual devices do not work unless pipewire's configuration files are changing to enable them.

**Comment by sim590** [comment]
I'm trying to figure out what config to change. According to [this forum](https://forum.endeavouros.com/t/cant-control-volume-of-virtual-pipewire-device/44099), I'd have to set `monitor.channel-volumes` to `true`, but this parameter seems to apply to some config that creates a virtual, but the virtual device `Easy Effects Source` is created by Easy Effects and I don't have control on the args it uses, do I? Do you have any insight as to how I could achieve what you suggested when you said "unless pipewire's configuration files are changing to enable them"?

**Comment by wwmm** [comment]
> I'm trying to figure out what config to change. According to [this forum](https://forum.endeavouros.com/t/cant-control-volume-of-virtual-pipewire-device/44099), I'd have to set `monitor.channel-volumes` to `true`, but this parameter seems to apply to some config that creates a virtual device, but the virtual device `Easy Effects Source` is created by Easy Effects and I don't have control on the args it uses, do I? Do you have any insight as to how I could achieve what you suggested when you said "unless pipewire's configuration files are changing to enable them"?

If I remember well there was a way to force pipewire to set `monitor.channel-volumes = true` globally. Doing it only at the moment the device was created was not the only way. It may be worth to ask about this at pipewire's page.

Dou you a have a reason so set these controls? Usually configuring the hardware volume and mute controls is enough.

The reason we do not force these controls to be enabled anymore is because they caused a lot of undesirable side effects for some users. Like the lack of sound because something  outside of easyeffects muted the device.

**Comment by sim590** [comment]
@wwmm: [this issue](https://github.com/Alexays/Waybar/issues/3712) should explain what I'm trying to achieve. For now, I'm trying to see if there's a way around on Waybar's side of things.

But also, I would have liked to have a simpler configuration in Sway for instance for handling my microphone's mute status without depending on the source name. For now, I'm forced to invoke:

```
pamixer --source alsa_input.usb-Blue_Blue_Snowflake-00.mono-fallback -t
```

instead of the simpler `exec pamixer --default-source -t`.

**Comment by wwmm** [comment]
> @wwmm: [this issue](https://github.com/Alexays/Waybar/issues/3712) should explain what I'm trying to achieve. For now, I'm trying to see if there's a way around on Waybar's side of things. I'm trying to have a visual feedback whenever my microphone is muted.
> 
> But also, I would have liked to have a simpler configuration in Sway for instance for handling my microphone's mute status without depending on the source name. For now, I'm forced to invoke:
> 
> ```
> pamixer --source alsa_input.usb-Blue_Blue_Snowflake-00.mono-fallback -t
> ```
> 
> instead of the simpler `exec pamixer --default-source -t`.

Never set EasyEffects virtual devices as default https://github.com/wwmm/easyeffects/discussions/1836#discussioncomment-10500670.  The application is designed assuming your hardware remains as default.  This may make it fail in some cases.

**Comment by sim590** [comment]
I guess it's working for me because I did set "use default input" to "no":
![image](https://github.com/user-attachments/assets/a66e13d4-201a-4c8f-b7ac-4dfce1bbdd6a)
In the thread you linked, you mention:
>For example when Process All Output Streams is enabled in our configuration window EasyEffects will move streams to its virtual output on its own.
Is there an equivalent for input streams? I'm really looking to make it so my games and specific apps which don't give you any option to change the device you want as microphone is gonna just use the right device, i.e. the device with the effects applied.

**Comment by wwmm** [comment]
>I guess it's working for me because I did set "use default input" to "no":

Maybe.

>Is there an equivalent for input streams?

When `Process All Input Streams` is enabled the same thing is done. All streams that are recording from the microphone being processed by EasyEffects will be moved to its virtual source. With the exception of the very rare cases where the app developer sets a flag that asks PipeWire to not allow their streams to be moved.
